### DISCUSSION TITLE
middleware foo from documentation stop working

### DISCUSSION THREAD

**Discussion by gayratv** [pmndrs/zustand#2422] [Bug report]
I'm studying the manual
docs/guides//typescript.md

Middleware that changes the store type:
https://github.com/pmndrs/zustand/blob/08fd86441710dc3e0791a7ae5a1461eb1232f84c/docs/guides/typescript.md?plain=1#L286

copy code from manual to TS playground :

[TS playground link](https://www.typescriptlang.org/play?ts=5.4.2#code/JYWwDg9gTgLgBAbwFB1XAxlApgQxlgGhTQGUY8sBhbPaItOM6LAWQFdyZoBJAEywB2MYADNgWKPTTtOhYqibYAgmGBEAvnBFQIIOAHIAXmwDO5Ab31IkMAJ5gscAOpRg+ADwAVOFgAe+CxM4CAAjACssdBgCOABVH39BXiDQiKiAPjgAXjgAeRA3LxiAayxbCBE4zIAyOOs7BzhKHDMiquy4bz8A5Li4AH5OuAAuOvr7RwAxCAgO93khqVQlJbgWMCDupKCAbUVWDlooPkFhMQkYtgFigQgAdwEAXR3HjpfVlnRNxMC4Pa5sDIjichKJxJI4Fcbvcni83o8iOkABQLESjMgUai4AFtHYAOgJ6xMMR2+hEM30MSUjwRay+6VWIRwUFGKyQAEpspkMfgsUc2kSSaTyRBKXBqTECXjPiZHulrPx0AAbZmOEAQXhsJWOIymcyWRALYBCCQiHDoRz7IEAkzuEhUzLIBgMEWjFxuLDuZqte3BcKRGAMxBaGasuDqeUMdRIaNAA)

and see error:
```
declare module 'zustand' {
>>>Error: All declarations of 'StoreMutators' must have identical type parameters.
    interface StoreMutators<S, A> {
        foo: Write<Cast<S, object>, { foo: A }>
    }
}
```
TS Playground show that StoreMutators declared as interface StoreMutators<S, A, S, A>
Yesterday this code was worked.
Webstorm alsso generate same error in project

May be you have some ideas?

**Comment by gayratv** [comment]
I'm insert react.ts and vanilla.ts from zustand to my local project and problem was gone.
I think problem inside npm package

**Reply by dai-shi** [reply]
Can you try different Zustand versions?

**Comment by gayratv** [comment]
@dai-shi 

I try 4.5.0 - same effect
I fork stackblitz project from #2175 and simply add file middleware.ts (from documentation) - and nothing doing else

[stackblitz ](https://stackblitz.com/edit/stackblitz-starters-uum8k7?file=app%2F_components%2FTestClientComponent.tsx,stores%2Fmiddleware%2Fmiddleware.ts)

and see many errors , for example
![Снимок экрана 2024-03-19 111053](https://github.com/pmndrs/zustand/assets/33682052/aebe19c6-d914-49cc-bfc7-068653d9cc6c)

**Comment by dai-shi** [comment]
> Yesterday this code was worked.

Hmm, I think we need to know what is changed.
